ddtrace/tracer/sampler.go

```diff
@@ -147,3 +154,225 @@ func (ps *prioritySampler) apply(spn *span) {
 	}
 	spn.SetTag(keySamplingPriorityRate, rate)
 }
+
+type rulesSampler struct {
+	rules   []SamplingRule
+	rate    float64
+	limiter *rate.Limiter
+	// "effective rate" calculations
+	mu           sync.Mutex
+	ts           int64
+	allowed      int
+	total        int
+	previousRate float64
+}
+
+func newRulesSampler(rules []SamplingRule) *rulesSampler {
+	rate := sampleRate()
+	return &rulesSampler{
+		rules:   samplingRules(rules),
+		rate:    rate,
+		limiter: newRateLimiter(rate),
+	}
+}
+
+// samplingRules validates the user-provided rules and returns an internal representation.
+// If the DD_TRACE_SAMPLING_RULES environment variable is set, then the rules from
+// tracer.WithSamplingRules are ignored.
+func samplingRules(rules []SamplingRule) []SamplingRule {
+	rulesFromEnv := os.Getenv("DD_TRACE_SAMPLING_RULES")
+	if rulesFromEnv != "" {
+		rules = rules[:0]
+		jsonRules := []struct {
+			Service   string      `json:"service"`
+			Operation string      `json:"operation"`
+			Rate      json.Number `json:"rate"`
+		}{}
+		err := json.Unmarshal([]byte(rulesFromEnv), &jsonRules)
+		if err != nil {
+			log.Warn("error parsing DD_TRACE_SAMPLING_RULES: %v", err)
+			return nil
+		}
+		for _, v := range jsonRules {
+			if v.Rate == "" {
+				log.Warn("error parsing rule: rate not provided")
+				continue
+			}
+			rate, err := v.Rate.Float64()
+			if err != nil {
+				log.Warn("error parsing rule: invalid rate: %v", err)
+				continue
+			}
+			switch {
+			case v.Service != "" && v.Operation != "":
+				rules = append(rules, ServiceOperationRule(v.Service, v.Operation, rate))
+			case v.Service != "":
+				rules = append(rules, ServiceRule(v.Service, rate))
+			case v.Operation != "":
+				rules = append(rules, OperationRule(v.Operation, rate))
+			}
+		}
+	}
+	validRules := make([]SamplingRule, 0, len(rules))
+	for _, v := range rules {
+		if !(v.Rate >= 0.0 && v.Rate <= 1.0) {
+			log.Warn("ignoring rule %+v: rate is out of range", v)
+			continue
+		}
+		validRules = append(validRules, v)
+	}
+	return validRules
+}
+
+func sampleRate() float64 {
+	rate := 0.0
```

Comment:
Can we remove this line and just return `0` everywhere? It's not used anywhere else other than returns. Or, if you wish, just do:
```suggestion
	const defaultRate := 0.0
```

Reply:
Changed to `const defaultRate`. Its value is used in the warnings that get logged, in addition to being returned.